File: source/downloads/code/LeetCode/Python/Fraction-to-Recurring-Decimal.py
class Solution(object):
    def fractionToDecimal(self, numerator, denominator):
        """
        :type numerator: int
        :type denominator: int
        :rtype: str
        """
        if not numerator:
            return "0"
        ret = ""
        if numerator < 0 or denominator < 0:
            ret = "-"
        if numerator <0 and denominator < 0:
            ret = ""
        
        p = abs(numerator)
        d = abs(denominator)

        ret += str(p/d)
        if p%d == 0:
            return ret
        else:
            ret += "."
        index = 0
        # hash[remain:index of frac]
        hash = dict()
        frac = ""

        # 22/7
        if p > d:
            p = p-d*(p/d)
        
        # get fraction part
        while p%d not in hash:
            remain = p%d
            if remain == 0:
                break
            # 1/3
            if p < d:
                p = 10*p
            frac += str(p/d)
            p = p-d*(p/d)
            hash[remain] = index
            index += 1
        # p can divide d
        if p%d == 0:
            return ret+frac
        
        startindex = hash[p%d]
        return ret+frac[0:startindex]+"("+frac[startindex:index]+")"

File: downloads/code/LeetCode/Java/Fraction-to-Recurring-Decimal.java
public class Solution {
    public String fractionToDecimal(int numerator, int denominator) {
	if(numerator == 0) return "0";
	StringBuilder ret = new StringBuilder();
	// +/-
	ret.append((numerator>0) ^ (denominator>0)? "-" : "");
	long num = Math.abs((long)numerator);
	long den = Math.abs((long)denominator);

	// Integer part
	ret.append(num/den);
	num %= den;
	// no fractional part
	if(num==0) return ret.toString();

	// fractional part
	ret.append(".");
	// map<remainer, place you need to insert "(">
	HashMap<Long, Integer> map = new HashMap<Long, Integer>();
	map.put(num, ret.length());

	while(num!=0){
	    num *= 10;
	    ret.append(num/den);
	    num %= den;
	    if(map.containsKey(num)){
		// find repeating part
		int index = map.get(num);
		ret.insert(index, "(");
		ret.append(")");
		break;
	    }
	    else
		map.put(num,ret.length());
	}
	return ret.toString();
    }
}
